src/boot/sysfail.c

```diff
@@ -3,10 +3,37 @@
 #include "sysfail.h"
 #include "util.h"
 
-SysFailType sysfail_check(void) {
+static bool firmware_update_is_failed(void) {
+        const EFI_SYSTEM_RESOURCE_TABLE *esrt_table;
+        const EFI_SYSTEM_RESOURCE_ENTRY *esrt_entries;
+
+        esrt_table = find_configuration_table(MAKE_GUID_PTR(EFI_SYSTEM_RESOURCE_TABLE));
+        if (!esrt_table)
+                return false;
+
+        esrt_entries = (const EFI_SYSTEM_RESOURCE_ENTRY*)((const uint8_t*)esrt_table + sizeof(EFI_SYSTEM_RESOURCE_TABLE));
+
+        FOREACH_ARRAY(esrt_entry, esrt_entries, esrt_table->FwResourceCount)
+                if (esrt_entry->FwType == ESRT_FW_TYPE_SYSTEMFIRMWARE)
+                        return esrt_entry->LastAttemptStatus != LAST_ATTEMPT_STATUS_SUCCESS;
+
+        return false;
+}
+
+SysFailType sysfail_check(SysFailConfig *sysfail_config) {
+        if (sysfail_config->check_firmware_update && firmware_update_is_failed())
+                return SYSFAIL_FIRMWARE_UPDATE;
+
         return SYSFAIL_NO_FAILURE;
 }
 
 const char16_t* sysfail_get_error_str(SysFailType fail_type) {
-        return NULL;
+        switch (fail_type) {
+        case SYSFAIL_NO_FAILURE:
+                return NULL;
+        case SYSFAIL_FIRMWARE_UPDATE:
+                return u"FirmwareUpdateFailed";
```

Comment:
in systemd we usually write enum values in lowercase with dashes. hence maybe return "firmware-update-failed" here.

Reply:
also, please make sure that the string and the C name are the same (except for case and separators, and a prefix). Hence maybe SYSFAIL_FIRMWARE_UPDATE_FAILURE and "firmware-updare-failure".